app/models/concerns/variant_stock.rb

```diff
@@ -40,9 +40,9 @@ def on_hand=(new_level)
 
   # Checks whether this variant is produced on demand.
   def on_demand
-    # A variant that has not been saved yet, doesn't have a stock item
+    # A variant that has not been saved yet or has been soft-deleted doesn't have a stock item
     #   This provides a default value for variant.on_demand using Spree::StockLocation.backorderable_default
-    return Spree::StockLocation.first.backorderable_default if stock_items.empty?
+    return Spree::StockLocation.first.backorderable_default if new_record? || deleted?
```

Comment:
nice catch!
Maybe there will be situations where the record isnt new, the record isnt deleted and still there is no stock_item... :boom: 
I'd keep both your conditions and maybe add the previous conditional on stock_items.empty? #defensive programming

Reply:
in any case, this needs to be unit tested in variant_stock_spec.

Reply:
The problem is, if it's not new or deleted and we call `stock_items.empty?`, we trigger a pointless query in this line that counts the variant's stock items to see if any exist. If it's not new or deleted it will always have one.

In cases like the cart where we're checking variant stock on lots of variants, if we have 30 items in the cart we will trigger this query 30 times. I'm pretty sure this happens elsewhere in the app as well, where we do a big N+1 query because of this `stock_items.empty?` check...

Reply:
I agree,
I am just not 100% sure about "If it's not new or deleted it will always have one."
If this is true then this code:
`return Spree::StockLocation.first.backorderable_default if new_record? || deleted? || stock_items.empty?`
will never run stock_items.empty?
That's why I'd add it to the end of the line to 100% sure.

Reply:
I see what you mean. But won't it always run the last bit in this example, when a variant is not new and not deleted?

Reply:
I've added a Bugsnag call for the case where there's a saved, non-deleted variant that does not have a `stock_item`, and ensured that the method will return what it previously would have returned in the case where `stock_items.empty? == true`.

Reply:
ok, I see. yes, it will call stock_items.empty? in those cases.
I wonder if stock_items.any? would be better
or even stock_items.first.nil? (which is what stock_item does later: stock_items.first)

I am ok with your proposed solution, I think we will see some bugsnags.
In this case I'd prefer to have a defensive code solution that does not rely on the state of the data.

Reply:
can you please add a unit test to variant_stock_spec?

Reply:
awesome, thanks!